src/V3Stats.cpp

```diff
@@ -202,6 +243,42 @@ class StatsVisitor final : public VNVisitor {
         allNodes(nodep);
         iterateChildrenConst(nodep);
     }
+    virtual void visit(AstActive* nodep) override {
+        m_countRegs = m_isPreOrder && nodep->hasClocked();
+        if (m_countRegs) iterateChildrenConst(nodep);
+    }
+    virtual void visit(AstVarRef* nodep) override {
+        AstVar* varp = nodep->varp();
+        if (m_countRegs && nodep->access().isWriteOrRW() && varp->isSignal()
+            && !varp->isUsedLoopIdx()) {
+            AstVarScope* varScopep = nodep->varScopep();
+            if (varScopep->user1()) {  // Already seen at least one assignment to this var
+                // Only count assignments on _different_ source lines as different since array
+                // indexed assignments may have been expanded into multiple AstVarRef by now
+                if (nodep->fileline() != (FileLine*)varScopep->user2p()) {
+                    varScopep->user1(varScopep->user1() + 1);
+                    varScopep->user2p((AstNode*)nodep->fileline());
```

Comment:
Instead of looking at the varref, I think instead have a visitor on the Var to determine if to insert it into your list.

The varref would set e.g. "if (!varp()->user2()) varp()->user2(nodep)" - so that iterating on all vars would see the varref that used it.

You'd then solve the visited-twice problem.  Also I think as presently a varref that crosses into a function or does something odd might get double counted.

Reply:
Need to get my head around this :)

Reply:
I think I get it, but since the var nodes are visited before the varrefs, how would I iterate the vars again without making another pass of the whole tree ?

Reply:
You're already making a vector of memories, so instead make that a vector (or std::set) on each var, then iterate it at the end.

Reply:
I considered that but wasn't sure if there was a better way. I've simplified this now